lib/Service/Avatar/BimiSource.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @author Gregor Mitzka <gregor.mitzka@gmail.com>
+ *
+ * Mail
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\Mail\Service\Avatar;
+
+use Exception;
+use Horde_Mail_Rfc822_Address;
+use OCP\Files\IMimeTypeDetector;
+use OCP\Http\Client\IClientService;
+
+class BimiSource implements IAvatarSource {
+
+	/** @var IClientService */
+	private $clientService;
+
+	/** @var IMimeTypeDetector */
+	private $mimeDetector;
+
+	/** @var DnsRecordService */
+	private $dnsRecordService;
+
+	public function __construct(
+		IClientService $clientService,
+		IMimeTypeDetector $mimeDetector,
+		DnsRecordService $dnsRecordService
+	) {
+		$this->clientService = $clientService;
+		$this->mimeDetector = $mimeDetector;
+		$this->dnsRecordService = $dnsRecordService;
+	}
+
+	/**
+	 * Does this source query external services?
+	 *
+	 * @return bool
+	 */
+	public function isExternal(): bool {
+		return true;
+	}
+
+	/**
+	 * @param string $email sender email address
+	 * @param AvatarFactory $factory
+	 * @return Avatar|null avatar URL if one can be found
+	 */
+	public function fetch(
+		string $email,
+		AvatarFactory $factory
+	): ?Avatar {
+		$horde = new Horde_Mail_Rfc822_Address($email);
+		// TODO: add support for other selectors
+		$domain = 'default._bimi.' . $horde->host;
```

Comment:
Does this handles IDN domains?

Reply:
Not sure, but it's the same way the `FaviconSource` handles hostnames. So either it works for both or for none.